field/TinyLang/Field/Evaluator.hs

```diff
@@ -144,22 +146,37 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
 evalBinOp Div x y = UniConst Field <$> divEval x y
 evalBinOp BAt x y = asIntegerEval x >>= fmap (UniConst Bool) . flip atEval y . fromIntegral
 
-evalStatementUni
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Statement f -> Expr f a -> m (UniConst f a)
-evalStatementUni env (ELet (UniVar _ var) def) rest = do
+evalProgramUni :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
+    => Env (SomeUniConst f) -> Program f -> m (Env (SomeUniConst f))
+evalProgramUni env = (flip execStateT) env . evalStatementsUni . unProgram
+
+evalStatementsUni :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
+    => Statements f -> StateT (Env (SomeUniConst f)) m ()
+evalStatementsUni = mapM_ evalStatementUni . unStatements
+
+evalStatementUni :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
+    => Statement f -> StateT  (Env (SomeUniConst f)) m ()
+evalStatementUni (ELet (UniVar _ var) def) = do
+    env <- get
     defR <- evalExprUni env def
-    evalExprUni (insertVar var (Some defR) env) rest
-evalStatementUni env (EAssert expr) rest = do
+    put $ insertVar var (Some defR) env
+evalStatementUni (EAssert expr) = do
+    env <- get
     exprR <- evalExpr env expr
-    if exprR
-        then evalExprUni env rest
-        else throwError $ AssertionFailedEvalError expr
+    unless exprR $ throwError $ AssertionFailedEvalError expr
+evalStatementUni (EFor (UniVar _ var) start end stmts)
+    | start <= end = mapM_ iteration [start .. end]
+    | otherwise    = modify' $ insertVar var (Some $ fromInteger actualEnd)
```

Comment:
We already know in that clause that `end > start`, so there isn't much point in doing `max start end`.

Reply:
It is a bit of defensive programming.  There is nothing preventing the user of writing the following programs:
```
for i = -1 to 0 do end;  assert i == 0;
```
```
for i = 0 to -1 do end;  assert i == 0;
```
```
for i = 0 to 0 do end;  assert i == 0;
```

This ensures that all of the above have defined semantics.

Reply:
Sure, but we already have two different clauses:

1. `start <= end`
2. `otherwise`, i.e. `start > end`

and so it's not necessary in the second clause to handle the `start <= end` case as it's already handled by the first clause.

Reply:
Yep, I can see it now.